src/data_distribution/first_touch_distribution.hpp

```diff
@@ -0,0 +1,131 @@
+		namespace atomic {
+			void _store_public_dir(const void* desired,
+				const std::size_t size, const std::size_t rank, const std::size_t disp);
+			void _store_local_dir(const std::size_t desired,
+				const std::size_t rank, const std::size_t disp);
+			void _load_local_dir(void* output_buffer,
+				const std::size_t rank, const std::size_t disp);
+			void _compare_exchange_dir(const void* desired, const void* expected, void* output_buffer,
+				const std::size_t size, const std::size_t rank, const std::size_t disp);
+		} // namespace atomic
+	} // namespace backend
+} // namespace argo
+
+#include "../backend/backend.hpp"
+
+/** @brief protects the owner directory */
+static pthread_mutex_t ownermutex = PTHREAD_MUTEX_INITIALIZER;
+
+namespace argo {
+	namespace data_distribution {
+		/**
+		 * @brief the first-touch data distribution
+		 * @details gives ownership of a page to the node that first touched it
+		 */
+		template<int instance>
+		class first_touch_distribution : public base_distribution<instance> {
+			private:
+				/**
+				 * @brief try and claim ownership of a page
+				 * @param addr address in the global address space
+				 * @return the homenode of addr
+				 */
+				static std::size_t first_touch (const std::size_t& addr);
+			public:
+				virtual node_id_t homenode (char* const ptr) {
+					std::size_t homenode;
+					const std::size_t addr = ptr - base_distribution<instance>::start_address;
+					const std::size_t index = 2 * (addr / granularity);
+					pthread_mutex_lock(&ownermutex);
+					argo::backend::atomic::_load_local_dir(&homenode, argo::backend::node_id(), index);
+					if (!homenode) { homenode = first_touch(addr); }
+					pthread_mutex_unlock(&ownermutex);
+
+					int n;
+					for(n = 0; n < base_distribution<instance>::nodes; n++) {
+						if((1UL << n) == homenode) {
+							homenode = n;
+						}
+					}
+
+					if(homenode >=(std::size_t)base_distribution<instance>::nodes) {
+						exit(EXIT_FAILURE);
+					}
+					return homenode;
+				}
+
+				virtual std::size_t local_offset (char* const ptr) {
+					std::size_t offset;
+					const std::size_t addr = ptr - base_distribution<instance>::start_address;
+					const std::size_t drift = addr % granularity;
+					const std::size_t index = 2 * (addr / granularity);
+					pthread_mutex_lock(&ownermutex);
+					argo::backend::atomic::_load_local_dir(&offset, argo::backend::node_id(), index+1);
+					offset += drift;
+					pthread_mutex_unlock(&ownermutex);
+
+					if(offset >=(std::size_t)base_distribution<instance>::size_per_node) {
+						exit(EXIT_FAILURE);
+					}
+					return offset;
+				}
+		};
+
+		template<int instance>
+		std::size_t first_touch_distribution<instance>::first_touch (const std::size_t& addr) {
+			/* variables for CAS */
+			node_id_t homenode;
+			std::size_t result;
+			constexpr std::size_t compare = 0;
+			const std::size_t id = 1 << argo::backend::node_id();
+			const std::size_t index = 2 * (addr / granularity);
```

Comment:
Unclear what this index is, could use a more descriptive name. Suggesting `owner_window_index` (if `ownerWindow` is renamed to `owner_window`).

Reply:
Changed as requested.